service_config.md

```diff
@@ -0,0 +1,194 @@
+ |  |   snippet1: |                      |  | 04-secret-<service>.conf         |
+ |  | 	[logger_root]                    |  |                                  |
+ |  | 	level=WARNING                    |  | [logger_root]                    |
+ |  | 	handlers=stdout                  |  | level=WARNING                    |
+ |  |   snippet2: |                      |  | handlers=stdout                  |
+ |  | 	##################               |  | ##################               |
+ |  | 	# Log Formatters #               |  | # Log Formatters #               |
+ |  | 	##################               |  | ##################               |
+ |  | 	[formatter_normal]               |  | [formatter_normal]               |
+ |  | 	format=(%(name)s): s%(message)s  |  | format=(%(name)s): s%(message)s  |
+ |  +------------------------------------+  | [formatters]                     |
+ |                                          | keys=normal                      |
++-------------------------+                 +----------------------------------+
+| apiVersion: v1          |                           |
+| kind: Secret            |                           |
+| metadata:               |                           |
+|   name: service-secret2 |                           |
+|   namespace: openstack  |        1. Mounted as Volume by Service Pod
+| stringData:             |
+|   snippet3: |           |        2. `kolla_set_configs && kolla_start`: sync
+|  [formatters]           |           the resulting secret as done for the other
+|  keys=normal            |           regular config files
++-------------------------+
+```
+
+As a result of this strategy, the service presents a layout similar to the
+following:
+
+```
+00-default.conf            => default configs generated by operator. This is stored in ConfigMap
+01-deployment-secrets.conf => default configs generated by operator, which contains credentials such as [database] connection . This is stored in Secret
+02-global.conf             => custom configs provided by users via top-level customServiceConfig. This is stored in ConfigMap
+03-service.conf            => custom configs provided by users via service level customServiceConfig. This is stored in ConfigMap
+04-secrets.conf            => custom configs provided by users via service level customServiceConfigSecrets, which contains credentials. This is stored in Secret and would not be present if no secrets are provided
+```
+
+
+The service will pass the `--config-dir` parameter to point to the
+`<service>.conf.d` directory where all the config files listed above are
+rendered.
+
+```
+"command": "/usr/bin/<service> --config-dir /etc/<service>/<service>.conf.d"
+```
+
+If this strategy is not available, the `init container` that executes a start
+script (e.g., `init.sh`) won't be removed, and the logic that generates the
+layout mentioned above will be implemented in the init script.
```

Comment:
I suggest that in this case we raise an upstream blueprint / RFE to add support for it.

Reply:
all service that use oslo.cofnig supprot 

/etc/<service>/<service>.conf.d by default.

so i think this would only be relevant for non OpenStack services
like ovs.

we may or may not want to expose configurabltiy of those services.

for example we currently don't plan to expose configuring libvirt unless we have a customer usecase that forces that. currently we do not have a usecase that requires it.

i agree with gibi if an OpenStack service cannot support config directories this should be addressed upstream in that service.

Reply:
I agree but I don't want to enforce services to implement this spec.
I think it's great that we can come up with something we share and agree on, but we can have exceptions, it shouldn't be a problem given that the limitation is in the service itself (not supporting --config-dir for instance).
A given service can take this, follow up upstream with the relevant people and team, and come back to this solution.
How does it sound?

Reply:
Sounds goo.